### PR TITLE
Zds 41/product template

### PR DESCRIPTION
**Pull request by admmasters**
For context this is the main page template to be used to contain and layout elements on the page.

Most of this is ported from signup, quopa and evva

Remaining Tasks
- [x] Template
- [x] Card
- [x] Header
- [x] Loading
- [x] Progress
- [x] Tests
- [x] Documentation

Tests to be added, but the api shape should be there - sub components hang off the ProductTemplate export:

```
ProductTemplate
ProductTemplate.Card
ProductTemplate.Header
ProductTemplate.Loading
ProductTemplate.Progress
```

### CODE REVIEW

File: src/components/templates/Product/Card.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  const { width = 0 } = useViewport();
+  return width > breakpoints.phone ? (
+    <Heading as="h2" size="h3">
+      {props.children}
+    </Heading>
+  ) : (
+    <Heading as="h2" size="h4">
+      {props.children}
+    </Heading>
+  );
```
Comment (raff-r) [inline src/components/templates/Product/Card.tsx:31]:
```suggestion
  return  (
    <Heading as="h2" size={width > breakpoints.phone ? "h3" : "h4"}>
      {props.children}
    </Heading>
  );
```

File: src/components/templates/Product/Card.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import React, { ReactNode } from 'react';
+import styled from 'styled-components';
+import classnames from 'classnames';
+
+import { Card, breakpoints, Heading, Text, colors, useViewport } from '../../../index';
+
+interface ProductTemplateCardProps {
+  notification?: ReactNode;
+  title?: ReactNode;
+  subtitle?: ReactNode;
```
Comment (raff-r) [inline src/components/templates/Product/Card.tsx:10]: Do we need to specify props for these, could they not be achieved by passing them in as part of `children` giving teams more flexibility

File: src/components/templates/Product/Template.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  subtitle?: string;
+  children: React.ReactNode;
+}
+
+const ProductTemplate = (props: ProductTemplateProps) => {
+  return (
+    <section data-automation="ZA.ProductTemplate">
+      <ProductTemplateHeader title={props.title} subtitle={props.subtitle} />
+      <ProductTemplateContainer> {props.children} </ProductTemplateContainer>
+      <ZopaFooter />
```
Comment (raff-r) [inline src/components/templates/Product/Template.tsx:21]: We could leave Footer out of this. The template is going to be used on a page by page basis and doesn't really make sense to have to re-render the footer multiple times.

File: src/components/templates/Product/Card.tsx (hunk 2 above)
Comment (admmasters) [reply src/components/templates/Product/Card.tsx:10]: Yeah thats fair

File: src/components/templates/Product/Template.tsx (hunk 3 above)
Comment (admmasters) [reply src/components/templates/Product/Template.tsx:21]: Makes sense - fair enough

File: src/components/templates/Product/Progress.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+interface ProductTemplateProgressProps {
+  nextStepUrl: string;
+  prevStepUrl?: string;
+  progress?: Pick<ProgressProps, 'currentStep' | 'totalSteps'>;
+  children: React.ReactNode;
+}
+
+const BackWrapper = styled.div`
+  ${minMedia.desktop`
```
Comment (JuliaHempel) [inline src/components/templates/Product/Progress.tsx:15]: I get this error `Uncaught TypeError: Cannot read property 'desktop' of undefined
    at Module../src/components/templates/Product/Progress.tsx (Progress.tsx:14)`

Comment (poteirard) [reply src/components/templates/Product/Progress.tsx:15]: I'm working on it

File: src/components/templates/Product/Card.tsx (hunk 2 above)
Comment (poteirard) [reply src/components/templates/Product/Card.tsx:10]: I think the point of the template page is to make it the less flexible as possible. If we start giving flexibility what's the point of the template?

File: src/components/templates/ProductTemplate/ProductTemplateLoading/ProductTemplateLoading.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+### Summary
+
+`<ProductTemplateLoading>`
+
+// TODO
```
Comment (poteirard) [inline src/components/templates/ProductTemplate/ProductTemplateLoading/ProductTemplateLoading.md:5]: I don't understand the purpose of this component

File: src/components/templates/ProductTemplate/index.tsx [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { ProductTemplateLoading } from './ProductTemplateLoading/ProductTemplateLoading';
+import { ProductTemplateProgress } from './ProductTemplateProgress/ProductTemplateProgress';
+import { ProductTemplateCard } from './ProductTemplateCard/ProductTemplateCard';
+import ProductTemplateComponent, { ProductTemplateProps } from './ProductTemplate/ProductTemplate';
+import React from 'react';
```
Comment (admmasters) [inline src/components/templates/ProductTemplate/index.tsx:5]: Think the standard in zrc is to have the package imports at the top - followed by the relative imports with a space separator

File: src/components/templates/ProductTemplate/index.tsx [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+interface ProductTemplateStatic {
+  Loading: typeof ProductTemplateLoading;
+  Progress: typeof ProductTemplateProgress;
+  Card: typeof ProductTemplateCard;
+}
+
+const ProductTemplate: ProductTemplateStatic & React.FC<ProductTemplateProps> = (props) => (
```
Comment (admmasters) [inline src/components/templates/ProductTemplate/index.tsx:13]: I've noticed this is a trend in this repo, but there seem to be quite a few opinions against using React.FC (which is why I used the other form) - https://fettblog.eu/typescript-react-why-i-dont-use-react-fc/

Comment (admmasters) [reply src/components/templates/ProductTemplate/index.tsx:13]: https://levelup.gitconnected.com/ultimate-react-component-patterns-with-typescript-2-8-82990c516935

File: src/components/templates/ProductTemplate/ProductTemplateHeader/ProductTemplateHeader.tsx [hunk 8]
```diff
@@ -1,7 +1,10 @@
 import React from 'react';
 import styled from 'styled-components';
-
-import { breakpoints, colors, Heading, Text, useViewport } from '../../../index';
+import { colors } from '../../../../constants/colors';
```
Comment (admmasters) [inline src/components/templates/ProductTemplate/ProductTemplateHeader/ProductTemplateHeader.tsx:3]: Spacing between the imports

File: src/components/templates/ProductTemplate/ProductTemplateLoading/ProductTemplateLoading.md (hunk 5 above)
Comment (admmasters) [reply src/components/templates/ProductTemplate/ProductTemplateLoading/ProductTemplateLoading.md:5]: It was a WIP - not sure its really needed - but it was originally there to represent a loading state for the template - which arguably we should have but don't at the moment.

Comment (admmasters) [reply src/components/templates/ProductTemplate/ProductTemplateLoading/ProductTemplateLoading.md:5]: At the moment it doesn't do anything so can kill it

File: src/components/templates/ProductTemplate/index.tsx (hunk 7 above)
Comment (poteirard) [reply src/components/templates/ProductTemplate/index.tsx:13]: I think we should stick one way of doing things for consistency. Maybe we can create an issue in our ZSC board so someone can have a look at it and change everything.

File: src/components/templates/ProductTemplate/ProductTemplateLoading/ProductTemplateLoading.md (hunk 5 above)
Comment (raff-r) [reply src/components/templates/ProductTemplate/ProductTemplateLoading/ProductTemplateLoading.md:5]: There are separate designs for loading state, there is a full page version and modal version

File: src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+### Summary
+
+`<ProductTemplateProgress>` the _<Progress />_ component with fixed styling.
```
Comment (raff-r) [inline src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.md:3]: adding `_<Progress />_` adds a funky progress bar in the docs
Screenshot 2020-09-21 at 11 14 38

File: src/components/templates/ProductTemplate/ProductTemplateCard/ProductTemplateCard.tsx [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import Text from '../../../atoms/Text/Text';
+import Card from '../../../organisms/Card/Card/Card';
+import { breakpoints } from '../../../../constants/breakpoints';
+import { colors } from '../../../../constants/colors';
+import { maxMedia } from '../../../../helpers/responsiveness';
+
+interface ProductTemplateCardProps {
+  notification?: ReactNode;
+  title?: ReactNode;
+  subtitle?: ReactNode;
```
Comment (raff-r) [inline src/components/templates/ProductTemplate/ProductTemplateCard/ProductTemplateCard.tsx:15]: Did we decide against removing these props in favour for just passing in `children`?

Comment (poteirard) [reply src/components/templates/ProductTemplate/ProductTemplateCard/ProductTemplateCard.tsx:15]: It's still an open discussion https://github.com/zopaUK/react-components/pull/590#discussion_r490264861

File: src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.md (hunk 9 above)
Comment (poteirard) [reply src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.md]: Fixed. Thanks!

File: src/components/templates/ProductTemplate/index.tsx (hunk 6 above)
Comment (poteirard) [reply src/components/templates/ProductTemplate/index.tsx:5]: We can add https://eslint.org/docs/2.0.0/rules/sort-imports in the future. To force that

File: src/components/templates/ProductTemplate/ProductTemplate/ProductTemplate.test.tsx [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+describe('<ProductTemplate />', () => {
+  it('renders with title and children but without subtitle', () => {
+    const { container } = render(<ProductTemplate title="ProductTemplate title">All the children</ProductTemplate>);
+    expect(container).toMatchSnapshot();
+  });
+
+  it('renders with title, children and subtitle', () => {
+    const { container } = render(<ProductTemplate title="ProductTemplate title">All the children</ProductTemplate>);
+    expect(container).toMatchSnapshot();
```
Comment (poteirard) [inline src/components/templates/ProductTemplate/ProductTemplate/ProductTemplate.test.tsx:13]: I don't need we need that amount of snapshots. I would suggest to test for the presence of the sub components and leave only one snapshot per component.

File: src/components/templates/Product/Card.tsx (hunk 2 above)
Comment (raff-r) [reply src/components/templates/Product/Card.tsx]: I agree, we should make it less flexible for layout, header etc but this for the content area. We should be able to put in any content we like. We already accept children here so I can see teams ignoring the title, subtitle and notification props and just passing in what content they want to show

Comment (poteirard) [reply src/components/templates/Product/Card.tsx]: Ahh yes, make sense for the content area. Agree

File: src/components/templates/ProductTemplate/index.tsx (hunk 7 above)
Comment (admmasters) [reply src/components/templates/ProductTemplate/index.tsx:12]: Fair enough - that makes complete sense

File: src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.tsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+import styled from 'styled-components';
+
+import Progress, { ProgressProps } from '../../../molecules/Progress/Progress';
+import { minMedia } from '../../../../helpers/responsiveness';
+
+interface ProductTemplateProgressProps {
+  nextStepUrl: string;
```
Comment (raff-r) [inline src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.tsx:8]: This isn't used. Can we remove?

File: src/components/templates/ProductTemplate/README.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+const ProductTemplateExample = () => (
+  <ProductTemplate title="Product title" subtitle="Product subtitle">
+    <ProductTemplate.Progress
+      prevStepUrl="https://zopa.com"
+      nextStepUrl="https://zopa.com"
+      progress={{ currentStep: 2, totalSteps: 4 }}
+    >
+      Body of the progress component
+    </ProductTemplate.Progress>
+    <ProductTemplate.Card title="Title of the card" subtitle="subtitule of the card">
```
Comment (raff-r) [inline src/components/templates/ProductTemplate/README.md:19]: title and subtitle props no longer exist on ProductTemplate.Card

File: src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.tsx [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  min-height: 30px;
+`;
+
+export function ProductTemplateProgress({ prevStepUrl, progress, children }: ProductTemplateProgressProps) {
+  if (!prevStepUrl || !progress) {
+    return null;
+  }
+
+  return (
+    <ProductTemplateProgressContainer data-automation="ZA.ProductTemplateProgress" className="l:mt-10 l:mb-4">
```
Comment (raff-r) [inline src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.tsx:33]: The spacing for the progress container seems a bit large on desktop and small on mobile. I've attached references to some designs in the Jira ticket https://jira.zopa/browse/ZDS-41

File: src/components/templates/ProductTemplate/ProductTemplateCard/ProductTemplateCard.tsx [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+import Card from '../../../organisms/Card/Card/Card';
+import { maxMedia } from '../../../../helpers/responsiveness';
+
+interface ProductTemplateCardProps {
+  className?: string;
+  children: React.ReactNode;
+}
+
+const StyledCard = styled(Card)`
+  ${maxMedia.phone`
```
Comment (raff-r) [inline src/components/templates/ProductTemplate/ProductTemplateCard/ProductTemplateCard.tsx:13]: When on mobile it should also remove drop shadow and have a border bottom

File: src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.tsx [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+      {prevStepUrl && <BackWrapper className="my-4 l:mt-0 l:mb-7 pl-2">{children}</BackWrapper>}
```
[the remark this answers is not in the document]
Comment (steedems) [reply src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.tsx:34]: nay chance we can have these elements decoupled from here as well? Gatsby and NextJS would probably need to use their own Link 🤔 

The usage I have in mind would be something like:

```
<ProductTemplate title="Product title" subtitle="Product subtitle">
    <ProductTemplate.Head>
        <ProductTemplate.Back>
            // Whatever flexible Link back button
        </ProductTemplate.Back>
        <ProductTemplate.Progress progress={{ currentStep: 2, totalSteps: 4 }}>
    </ProductTemplate.Head>
    <ProductTemplate.Card>
        This is the body of the card
    </ProductTemplate.Card>
</ProductTemplate>
```

Or maybe we can have a default Back and it's up to the product to use it as is or implement a custom one

```
<ProductTemplate title="Product title" subtitle="Product subtitle">
    <ProductTemplate.Head>
        <ProductTemplate.Back back button props here />
        <ProductTemplate.Progress progress={{ currentStep: 2, totalSteps: 4 }}>
    </ProductTemplate.Head>
    <ProductTemplate.Card>
        This is the body of the card
    </ProductTemplate.Card>
</ProductTemplate>
```

[the remark this answers is not in the document]
Comment (JuliaHempel) [reply src/components/templates/ProductTemplate/ProductTemplateProgress/ProductTemplateProgress.tsx]: I've made it so you can optionally send in `prevStep` and if it's a string I put it in the standard back button and if it's a ReactNod I just place it in the component. 

```  
<ProductTemplate
    title="Product title"
    subtitle="Product subtitle"
    prevStep="prevStep"
    progress={{ currentStep: 2, totalSteps: 4 }}
  >
    <ProductTemplate.Card>This is the body of the card</ProductTemplate.Card>
  </ProductTemplate>
```
or as a node
```  
<ProductTemplate
    title="Product title"
    subtitle="Product subtitle"
    prevStep={<>custom prevStep link</>}
    progress={{ currentStep: 2, totalSteps: 4 }}
  >
    <ProductTemplate.Card>This is the body of the card</ProductTemplate.Card>
  </ProductTemplate>
```

**Review by raff-r — APPROVED** [review]
Great effort everyone! thank you! 🎉

**Pull request merged**